{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os, subprocess, gzip, re"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DF_files = pd.read_csv('raw_data_info.csv',index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def gunzip(gz,out_dir):\n",
    "    basename = os.path.split(gz)[1][:-3]\n",
    "    result = os.path.join(out_dir,basename)\n",
    "    with gzip.open(gz,'rb') as f:\n",
    "        with open(result,'w') as f2:\n",
    "            f2.write(f.read())\n",
    "    return result\n",
    "\n",
    "def get_alignment_score(out_dir):\n",
    "    filename = os.path.join(out_dir,'bowtie_align.txt')\n",
    "    with open(filename,'r') as f:\n",
    "        result = f.readlines()[1]\n",
    "    match = re.search('\\([\\d\\.]*%\\)',result)\n",
    "    return float(result[match.start()+1:match.end()-2])\n",
    "\n",
    "def process_data(project,condition,rep,R1,R2,cores=8,verbose=False):\n",
    "        # Set base directory\n",
    "        ROOT_DIR = '/home/anand/Dropbox/Projects/biggdata_analysis/data/processed_data/'\n",
    "\n",
    "        ### Create processed file subdirectories ###\n",
    "\n",
    "        out_dir = os.path.join(ROOT_DIR,project,condition,str(rep))\n",
    "        basename = '__'.join([project,condition,str(rep)])\n",
    "        \n",
    "        # Quit if file already exists\n",
    "        if os.path.isfile(os.path.join(out_dir,basename+'.bam')):\n",
    "            score = get_alignment_score(out_dir)\n",
    "            return os.path.join(out_dir,basename+'.bam'),score\n",
    "        \n",
    "        if verbose:\n",
    "            print 'Processing %s -> %s -> %s'%(project,condition,rep)\n",
    "        \n",
    "        # Create directories if they don't already exist\n",
    "        if not os.path.isdir(out_dir):\n",
    "            if verbose:\n",
    "                print 'Creating directories %s'%out_dir\n",
    "            os.makedirs(out_dir)\n",
    "\n",
    "        ### Unzip fastq files ###\n",
    "\n",
    "        r1_files = []\n",
    "        r2_files = []\n",
    "        for fastq in R1.split(','):\n",
    "            if fastq.endswith('.gz'):\n",
    "                if verbose:\n",
    "                    print 'Unzipping file: %s'%fastq\n",
    "                r1_files.append(gunzip(fastq,out_dir))\n",
    "            else:\n",
    "                r1_files.append(fastq)\n",
    "\n",
    "        for fastq in R2.split(','):\n",
    "            if fastq.endswith('.gz'):\n",
    "                if verbose:\n",
    "                    print 'Unzipping file: %s'%fastq\n",
    "                r2_files.append(gunzip(fastq,out_dir))\n",
    "            else:\n",
    "                r2_files.append(fastq)\n",
    "\n",
    "        ### Run Bowtie Aligner ###\n",
    "\n",
    "        options = ['-X','1000','-n','2','-p',str(cores),'-3','3','-S']\n",
    "        bowtie_ref = '/home/anand/bin/bowtie-1.1.2/indexes/NC_000913.3'\n",
    "        bowtie_out = os.path.join(out_dir,basename+'.sam')\n",
    "        bowtie_err = os.path.join(out_dir,'bowtie_align.txt')\n",
    "        \n",
    "        if verbose:\n",
    "            print 'Running bowtie aligner...'\n",
    "        \n",
    "        with open(bowtie_out,'w') as out:\n",
    "            with open(bowtie_err,'w') as err:\n",
    "                subprocess.call(['bowtie']+options+['-1',','.join(r1_files),\n",
    "                             '-2',','.join(r2_files),bowtie_ref],\n",
    "                            stdout=out,stderr=err)\n",
    "\n",
    "        ### Post-process files ###\n",
    "\n",
    "        unsorted_bam = os.path.join(out_dir,basename+'.unsorted.bam')\n",
    "        sorted_bam = os.path.join(out_dir,basename+'.bam')\n",
    "        if verbose:\n",
    "            print 'Converting to BAM...'\n",
    "        subprocess.call(['samtools','view','-bS',bowtie_out,'-o',unsorted_bam])\n",
    "        if verbose:\n",
    "            print 'Sorting BAM file...'\n",
    "        subprocess.call(['samtools','sort',unsorted_bam,'-o',sorted_bam])\n",
    "\n",
    "        ### Clear all non-bam files ###\n",
    "        if verbose:\n",
    "            print 'Cleaning up...'\n",
    "        for f in os.listdir(out_dir):\n",
    "            if f.endswith('fastq'):\n",
    "                os.remove(os.path.join(out_dir,f))\n",
    "        os.remove(bowtie_out)\n",
    "        os.remove(unsorted_bam)\n",
    "        \n",
    "        ### Find alignment score ###\n",
    "        score = get_alignment_score(out_dir)\n",
    "\n",
    "        ### Add BAM file and alignment value to replicate ###\n",
    "        return sorted_bam,score"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# QC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of conditions: 114\n",
      "Number of unique conditions: 114\n",
      "Number of unique metadata files: 87\n"
     ]
    }
   ],
   "source": [
    "print 'Number of conditions: %d'%len(DF_files[['project','condition']].drop_duplicates())\n",
    "print 'Number of unique conditions: %d'%len(DF_files[['project','condition','metadata']].drop_duplicates())\n",
    "print 'Number of unique metadata files: %d'%len(DF_files.metadata.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Replicates: 174\n",
      "Number of unique R1 files: 174\n",
      "Number of unique R2 lists: 174\n"
     ]
    }
   ],
   "source": [
    "print 'Number of Replicates: %d'%len(DF_files[['project','condition','rep']].drop_duplicates())\n",
    "all_R1 = [r1.split(',') for r1 in DF_files.R1.values]\n",
    "all_R2 = [r1.split(',') for r1 in DF_files.R2.values]\n",
    "print 'Number of unique R1 files: %d'%len(DF_files.R1.unique())\n",
    "print 'Number of unique R2 lists: %d'%len(DF_files.R2.unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Assertions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Check that there are no duplicate R1/R2 files within each replicate\n",
    "assert all([len(r1.split(',')) == len(set(r1.split(','))) for r1 in DF_files.R1.unique()])\n",
    "assert all([len(r2.split(',')) == len(set(r2.split(','))) for r2 in DF_files.R2.unique()])\n",
    "\n",
    "# Ensure all R1/R2 files have R1/R2\n",
    "assert all([('R1' in r1) or ('_1.fastq' in r1) for f in DF_files.R1.unique() for r1 in f.split(',')])\n",
    "assert all([('R2' in r2) or ('_2.fastq' in r2) for f in DF_files.R2.unique() for r2 in f.split(',')])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process Reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for i,row in DF_files.iterrows():\n",
    "    bam,score = process_data(row.project,row.condition,row.rep,row.R1,row.R2,cores=4,verbose=True)\n",
    "    DF_files.loc[i,'BAM'] = bam\n",
    "    DF_files.loc[i,'alignment'] = score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DF_files.to_csv('raw_data_info.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project</th>\n",
       "      <th>condition</th>\n",
       "      <th>rep</th>\n",
       "      <th>metadata</th>\n",
       "      <th>R1</th>\n",
       "      <th>R2</th>\n",
       "      <th>BAM</th>\n",
       "      <th>alignment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>gadewx</td>\n",
       "      <td>delgadx_ph5</td>\n",
       "      <td>1</td>\n",
       "      <td>/media/nucleoid/raw_data/dhkim/gadewx/2016-02-...</td>\n",
       "      <td>/media/nucleoid/raw_data/dhkim/gadewx/2016-02-...</td>\n",
       "      <td>/media/nucleoid/raw_data/dhkim/gadewx/2016-02-...</td>\n",
       "      <td>/home/anand/Dropbox/Projects/biggdata_analysis...</td>\n",
       "      <td>46.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>gadewx</td>\n",
       "      <td>delgadx_ph5</td>\n",
       "      <td>2</td>\n",
       "      <td>/media/nucleoid/raw_data/dhkim/gadewx/2016-02-...</td>\n",
       "      <td>/media/nucleoid/raw_data/dhkim/gadewx/2016-02-...</td>\n",
       "      <td>/media/nucleoid/raw_data/dhkim/gadewx/2016-02-...</td>\n",
       "      <td>/home/anand/Dropbox/Projects/biggdata_analysis...</td>\n",
       "      <td>78.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>me_param</td>\n",
       "      <td>bw25113_gly</td>\n",
       "      <td>2</td>\n",
       "      <td>/media/nucleoid/raw_data/dhkim/me_param/2016-0...</td>\n",
       "      <td>/media/nucleoid/raw_data/dhkim/me_param/2016-0...</td>\n",
       "      <td>/media/nucleoid/raw_data/dhkim/me_param/2016-0...</td>\n",
       "      <td>/home/anand/Dropbox/Projects/biggdata_analysis...</td>\n",
       "      <td>82.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>minspan</td>\n",
       "      <td>bw25113_ade</td>\n",
       "      <td>1</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/SRR9...</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/SRR9...</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/SRR9...</td>\n",
       "      <td>/home/anand/Dropbox/Projects/biggdata_analysis...</td>\n",
       "      <td>78.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>minspan</td>\n",
       "      <td>bw25113_trp</td>\n",
       "      <td>1</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/SRR9...</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/SRR9...</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/SRR9...</td>\n",
       "      <td>/home/anand/Dropbox/Projects/biggdata_analysis...</td>\n",
       "      <td>78.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>minspan</td>\n",
       "      <td>bw25113_delnac_ade</td>\n",
       "      <td>1</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/SRR9...</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/SRR9...</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/SRR9...</td>\n",
       "      <td>/home/anand/Dropbox/Projects/biggdata_analysis...</td>\n",
       "      <td>81.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>minspan</td>\n",
       "      <td>bw25113_delcra_glc</td>\n",
       "      <td>1</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/SRR9...</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/SRR9...</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/SRR9...</td>\n",
       "      <td>/home/anand/Dropbox/Projects/biggdata_analysis...</td>\n",
       "      <td>84.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>minspan</td>\n",
       "      <td>bw25113_delcra_trp</td>\n",
       "      <td>1</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/SRR9...</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/SRR9...</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/SRR9...</td>\n",
       "      <td>/home/anand/Dropbox/Projects/biggdata_analysis...</td>\n",
       "      <td>76.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>minspan</td>\n",
       "      <td>bw25113_delmntr_anaero</td>\n",
       "      <td>1</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/SRR9...</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/SRR9...</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/SRR9...</td>\n",
       "      <td>/home/anand/Dropbox/Projects/biggdata_analysis...</td>\n",
       "      <td>67.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>minspan</td>\n",
       "      <td>bw25113_glc_anaero</td>\n",
       "      <td>1</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/wt_a...</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/wt_a...</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/wt_a...</td>\n",
       "      <td>/home/anand/Dropbox/Projects/biggdata_analysis...</td>\n",
       "      <td>65.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>minspan</td>\n",
       "      <td>bw25113_glc_anaero</td>\n",
       "      <td>2</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/wt_a...</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/wt_a...</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/wt_a...</td>\n",
       "      <td>/home/anand/Dropbox/Projects/biggdata_analysis...</td>\n",
       "      <td>83.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>minspan</td>\n",
       "      <td>bw25113_glc_anaero</td>\n",
       "      <td>3</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/wt_a...</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/wt_a...</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/minspan/wt_a...</td>\n",
       "      <td>/home/anand/Dropbox/Projects/biggdata_analysis...</td>\n",
       "      <td>72.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>glu</td>\n",
       "      <td>glu_ale7</td>\n",
       "      <td>1</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/glc_ale/ALE7...</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/glc_ale/ALE7...</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/glc_ale/ALE7...</td>\n",
       "      <td>/home/anand/Dropbox/Projects/biggdata_analysis...</td>\n",
       "      <td>25.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>rpoB</td>\n",
       "      <td>rpoBE546V_glc</td>\n",
       "      <td>1</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/rnap/rpoBE54...</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/rnap/rpoBE54...</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/rnap/rpoBE54...</td>\n",
       "      <td>/home/anand/Dropbox/Projects/biggdata_analysis...</td>\n",
       "      <td>87.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>c13</td>\n",
       "      <td>c13_1_133</td>\n",
       "      <td>1</td>\n",
       "      <td>/media/nucleoid/raw_data/troy_sandberg/SSW_RNA...</td>\n",
       "      <td>/media/nucleoid/raw_data/troy_sandberg/SSW_RNA...</td>\n",
       "      <td>/media/nucleoid/raw_data/troy_sandberg/SSW_RNA...</td>\n",
       "      <td>/home/anand/Dropbox/Projects/biggdata_analysis...</td>\n",
       "      <td>55.43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>c13</td>\n",
       "      <td>c13_2_134</td>\n",
       "      <td>1</td>\n",
       "      <td>/media/nucleoid/raw_data/troy_sandberg/SSW_RNA...</td>\n",
       "      <td>/media/nucleoid/raw_data/troy_sandberg/SSW_RNA...</td>\n",
       "      <td>/media/nucleoid/raw_data/troy_sandberg/SSW_RNA...</td>\n",
       "      <td>/home/anand/Dropbox/Projects/biggdata_analysis...</td>\n",
       "      <td>60.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>c13</td>\n",
       "      <td>c13_6_133</td>\n",
       "      <td>1</td>\n",
       "      <td>/media/nucleoid/raw_data/troy_sandberg/SSW_RNA...</td>\n",
       "      <td>/media/nucleoid/raw_data/troy_sandberg/SSW_RNA...</td>\n",
       "      <td>/media/nucleoid/raw_data/troy_sandberg/SSW_RNA...</td>\n",
       "      <td>/home/anand/Dropbox/Projects/biggdata_analysis...</td>\n",
       "      <td>59.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>ssw_glu_ac</td>\n",
       "      <td>ac_9_92_3</td>\n",
       "      <td>1</td>\n",
       "      <td>/media/nucleoid/raw_data/troy_sandberg/SSW_RNA...</td>\n",
       "      <td>/media/nucleoid/raw_data/troy_sandberg/SSW_RNA...</td>\n",
       "      <td>/media/nucleoid/raw_data/troy_sandberg/SSW_RNA...</td>\n",
       "      <td>/home/anand/Dropbox/Projects/biggdata_analysis...</td>\n",
       "      <td>53.34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>ssw_glu_ac</td>\n",
       "      <td>glu_9_92_3</td>\n",
       "      <td>1</td>\n",
       "      <td>/media/nucleoid/raw_data/troy_sandberg/SSW_RNA...</td>\n",
       "      <td>/media/nucleoid/raw_data/troy_sandberg/SSW_RNA...</td>\n",
       "      <td>/media/nucleoid/raw_data/troy_sandberg/SSW_RNA...</td>\n",
       "      <td>/home/anand/Dropbox/Projects/biggdata_analysis...</td>\n",
       "      <td>56.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>gly</td>\n",
       "      <td>gly_ale1</td>\n",
       "      <td>1</td>\n",
       "      <td>/media/nucleoid/raw_data/troy_sandberg/SSW_RNA...</td>\n",
       "      <td>/media/nucleoid/raw_data/troy_sandberg/SSW_RNA...</td>\n",
       "      <td>/media/nucleoid/raw_data/troy_sandberg/SSW_RNA...</td>\n",
       "      <td>/home/anand/Dropbox/Projects/biggdata_analysis...</td>\n",
       "      <td>58.83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>gly</td>\n",
       "      <td>gly_ale2</td>\n",
       "      <td>1</td>\n",
       "      <td>/media/nucleoid/raw_data/troy_sandberg/SSW_RNA...</td>\n",
       "      <td>/media/nucleoid/raw_data/troy_sandberg/SSW_RNA...</td>\n",
       "      <td>/media/nucleoid/raw_data/troy_sandberg/SSW_RNA...</td>\n",
       "      <td>/home/anand/Dropbox/Projects/biggdata_analysis...</td>\n",
       "      <td>57.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>misc</td>\n",
       "      <td>unknown_delcra</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/misc/cra_S2_...</td>\n",
       "      <td>/media/nucleoid/raw_data/avsastry/misc/cra_S2_...</td>\n",
       "      <td>/home/anand/Dropbox/Projects/biggdata_analysis...</td>\n",
       "      <td>88.78</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        project               condition  rep  \\\n",
       "10       gadewx             delgadx_ph5    1   \n",
       "11       gadewx             delgadx_ph5    2   \n",
       "45     me_param             bw25113_gly    2   \n",
       "79      minspan             bw25113_ade    1   \n",
       "80      minspan             bw25113_trp    1   \n",
       "82      minspan      bw25113_delnac_ade    1   \n",
       "83      minspan      bw25113_delcra_glc    1   \n",
       "84      minspan      bw25113_delcra_trp    1   \n",
       "86      minspan  bw25113_delmntr_anaero    1   \n",
       "90      minspan      bw25113_glc_anaero    1   \n",
       "91      minspan      bw25113_glc_anaero    2   \n",
       "92      minspan      bw25113_glc_anaero    3   \n",
       "100         glu                glu_ale7    1   \n",
       "106        rpoB           rpoBE546V_glc    1   \n",
       "118         c13               c13_1_133    1   \n",
       "119         c13               c13_2_134    1   \n",
       "120         c13               c13_6_133    1   \n",
       "127  ssw_glu_ac               ac_9_92_3    1   \n",
       "128  ssw_glu_ac              glu_9_92_3    1   \n",
       "131         gly                gly_ale1    1   \n",
       "132         gly                gly_ale2    1   \n",
       "138        misc          unknown_delcra    1   \n",
       "\n",
       "                                              metadata  \\\n",
       "10   /media/nucleoid/raw_data/dhkim/gadewx/2016-02-...   \n",
       "11   /media/nucleoid/raw_data/dhkim/gadewx/2016-02-...   \n",
       "45   /media/nucleoid/raw_data/dhkim/me_param/2016-0...   \n",
       "79   /media/nucleoid/raw_data/avsastry/minspan/SRR9...   \n",
       "80   /media/nucleoid/raw_data/avsastry/minspan/SRR9...   \n",
       "82   /media/nucleoid/raw_data/avsastry/minspan/SRR9...   \n",
       "83   /media/nucleoid/raw_data/avsastry/minspan/SRR9...   \n",
       "84   /media/nucleoid/raw_data/avsastry/minspan/SRR9...   \n",
       "86   /media/nucleoid/raw_data/avsastry/minspan/SRR9...   \n",
       "90   /media/nucleoid/raw_data/avsastry/minspan/wt_a...   \n",
       "91   /media/nucleoid/raw_data/avsastry/minspan/wt_a...   \n",
       "92   /media/nucleoid/raw_data/avsastry/minspan/wt_a...   \n",
       "100  /media/nucleoid/raw_data/avsastry/glc_ale/ALE7...   \n",
       "106  /media/nucleoid/raw_data/avsastry/rnap/rpoBE54...   \n",
       "118  /media/nucleoid/raw_data/troy_sandberg/SSW_RNA...   \n",
       "119  /media/nucleoid/raw_data/troy_sandberg/SSW_RNA...   \n",
       "120  /media/nucleoid/raw_data/troy_sandberg/SSW_RNA...   \n",
       "127  /media/nucleoid/raw_data/troy_sandberg/SSW_RNA...   \n",
       "128  /media/nucleoid/raw_data/troy_sandberg/SSW_RNA...   \n",
       "131  /media/nucleoid/raw_data/troy_sandberg/SSW_RNA...   \n",
       "132  /media/nucleoid/raw_data/troy_sandberg/SSW_RNA...   \n",
       "138                                                NaN   \n",
       "\n",
       "                                                    R1  \\\n",
       "10   /media/nucleoid/raw_data/dhkim/gadewx/2016-02-...   \n",
       "11   /media/nucleoid/raw_data/dhkim/gadewx/2016-02-...   \n",
       "45   /media/nucleoid/raw_data/dhkim/me_param/2016-0...   \n",
       "79   /media/nucleoid/raw_data/avsastry/minspan/SRR9...   \n",
       "80   /media/nucleoid/raw_data/avsastry/minspan/SRR9...   \n",
       "82   /media/nucleoid/raw_data/avsastry/minspan/SRR9...   \n",
       "83   /media/nucleoid/raw_data/avsastry/minspan/SRR9...   \n",
       "84   /media/nucleoid/raw_data/avsastry/minspan/SRR9...   \n",
       "86   /media/nucleoid/raw_data/avsastry/minspan/SRR9...   \n",
       "90   /media/nucleoid/raw_data/avsastry/minspan/wt_a...   \n",
       "91   /media/nucleoid/raw_data/avsastry/minspan/wt_a...   \n",
       "92   /media/nucleoid/raw_data/avsastry/minspan/wt_a...   \n",
       "100  /media/nucleoid/raw_data/avsastry/glc_ale/ALE7...   \n",
       "106  /media/nucleoid/raw_data/avsastry/rnap/rpoBE54...   \n",
       "118  /media/nucleoid/raw_data/troy_sandberg/SSW_RNA...   \n",
       "119  /media/nucleoid/raw_data/troy_sandberg/SSW_RNA...   \n",
       "120  /media/nucleoid/raw_data/troy_sandberg/SSW_RNA...   \n",
       "127  /media/nucleoid/raw_data/troy_sandberg/SSW_RNA...   \n",
       "128  /media/nucleoid/raw_data/troy_sandberg/SSW_RNA...   \n",
       "131  /media/nucleoid/raw_data/troy_sandberg/SSW_RNA...   \n",
       "132  /media/nucleoid/raw_data/troy_sandberg/SSW_RNA...   \n",
       "138  /media/nucleoid/raw_data/avsastry/misc/cra_S2_...   \n",
       "\n",
       "                                                    R2  \\\n",
       "10   /media/nucleoid/raw_data/dhkim/gadewx/2016-02-...   \n",
       "11   /media/nucleoid/raw_data/dhkim/gadewx/2016-02-...   \n",
       "45   /media/nucleoid/raw_data/dhkim/me_param/2016-0...   \n",
       "79   /media/nucleoid/raw_data/avsastry/minspan/SRR9...   \n",
       "80   /media/nucleoid/raw_data/avsastry/minspan/SRR9...   \n",
       "82   /media/nucleoid/raw_data/avsastry/minspan/SRR9...   \n",
       "83   /media/nucleoid/raw_data/avsastry/minspan/SRR9...   \n",
       "84   /media/nucleoid/raw_data/avsastry/minspan/SRR9...   \n",
       "86   /media/nucleoid/raw_data/avsastry/minspan/SRR9...   \n",
       "90   /media/nucleoid/raw_data/avsastry/minspan/wt_a...   \n",
       "91   /media/nucleoid/raw_data/avsastry/minspan/wt_a...   \n",
       "92   /media/nucleoid/raw_data/avsastry/minspan/wt_a...   \n",
       "100  /media/nucleoid/raw_data/avsastry/glc_ale/ALE7...   \n",
       "106  /media/nucleoid/raw_data/avsastry/rnap/rpoBE54...   \n",
       "118  /media/nucleoid/raw_data/troy_sandberg/SSW_RNA...   \n",
       "119  /media/nucleoid/raw_data/troy_sandberg/SSW_RNA...   \n",
       "120  /media/nucleoid/raw_data/troy_sandberg/SSW_RNA...   \n",
       "127  /media/nucleoid/raw_data/troy_sandberg/SSW_RNA...   \n",
       "128  /media/nucleoid/raw_data/troy_sandberg/SSW_RNA...   \n",
       "131  /media/nucleoid/raw_data/troy_sandberg/SSW_RNA...   \n",
       "132  /media/nucleoid/raw_data/troy_sandberg/SSW_RNA...   \n",
       "138  /media/nucleoid/raw_data/avsastry/misc/cra_S2_...   \n",
       "\n",
       "                                                   BAM  alignment  \n",
       "10   /home/anand/Dropbox/Projects/biggdata_analysis...      46.76  \n",
       "11   /home/anand/Dropbox/Projects/biggdata_analysis...      78.31  \n",
       "45   /home/anand/Dropbox/Projects/biggdata_analysis...      82.97  \n",
       "79   /home/anand/Dropbox/Projects/biggdata_analysis...      78.38  \n",
       "80   /home/anand/Dropbox/Projects/biggdata_analysis...      78.60  \n",
       "82   /home/anand/Dropbox/Projects/biggdata_analysis...      81.84  \n",
       "83   /home/anand/Dropbox/Projects/biggdata_analysis...      84.03  \n",
       "84   /home/anand/Dropbox/Projects/biggdata_analysis...      76.33  \n",
       "86   /home/anand/Dropbox/Projects/biggdata_analysis...      67.99  \n",
       "90   /home/anand/Dropbox/Projects/biggdata_analysis...      65.51  \n",
       "91   /home/anand/Dropbox/Projects/biggdata_analysis...      83.01  \n",
       "92   /home/anand/Dropbox/Projects/biggdata_analysis...      72.08  \n",
       "100  /home/anand/Dropbox/Projects/biggdata_analysis...      25.60  \n",
       "106  /home/anand/Dropbox/Projects/biggdata_analysis...      87.63  \n",
       "118  /home/anand/Dropbox/Projects/biggdata_analysis...      55.43  \n",
       "119  /home/anand/Dropbox/Projects/biggdata_analysis...      60.57  \n",
       "120  /home/anand/Dropbox/Projects/biggdata_analysis...      59.60  \n",
       "127  /home/anand/Dropbox/Projects/biggdata_analysis...      53.34  \n",
       "128  /home/anand/Dropbox/Projects/biggdata_analysis...      56.96  \n",
       "131  /home/anand/Dropbox/Projects/biggdata_analysis...      58.83  \n",
       "132  /home/anand/Dropbox/Projects/biggdata_analysis...      57.67  \n",
       "138  /home/anand/Dropbox/Projects/biggdata_analysis...      88.78  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_files[DF_files.alignment < 90]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## STOP HERE - Load Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "IOError",
     "evalue": "File /media/nucleoid/raw_data/troy_sandberg/SSW_RNAseq/1_133_S28/2016-10-01_RNA-seq.csv does not exist",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIOError\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-18-76a2d1685901>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mDF_metadata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mmetadata\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mDF_files\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmetadata\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m     \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmetadata\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mindex_col\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mheader\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m     \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'metadata'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmetadata\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mDF_metadata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconcat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mDF_metadata\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtranspose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/anand/.virtualenvs/main/local/lib/python2.7/site-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36mparser_f\u001b[0;34m(filepath_or_buffer, sep, dialect, compression, doublequote, escapechar, quotechar, quoting, skipinitialspace, lineterminator, header, index_col, names, prefix, skiprows, skipfooter, skip_footer, na_values, true_values, false_values, delimiter, converters, dtype, usecols, engine, delim_whitespace, as_recarray, na_filter, compact_ints, use_unsigned, low_memory, buffer_lines, warn_bad_lines, error_bad_lines, keep_default_na, thousands, comment, decimal, parse_dates, keep_date_col, dayfirst, date_parser, memory_map, float_precision, nrows, iterator, chunksize, verbose, encoding, squeeze, mangle_dupe_cols, tupleize_cols, infer_datetime_format, skip_blank_lines)\u001b[0m\n\u001b[1;32m    496\u001b[0m                     skip_blank_lines=skip_blank_lines)\n\u001b[1;32m    497\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 498\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    499\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    500\u001b[0m     \u001b[0mparser_f\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/anand/.virtualenvs/main/local/lib/python2.7/site-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    273\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    274\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 275\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    276\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    277\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mnrows\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mNone\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mchunksize\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/anand/.virtualenvs/main/local/lib/python2.7/site-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    588\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'has_index_names'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'has_index_names'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    589\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 590\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    591\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    592\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_get_options_with_defaults\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/anand/.virtualenvs/main/local/lib/python2.7/site-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, engine)\u001b[0m\n\u001b[1;32m    729\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'c'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    730\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'c'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 731\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mCParserWrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    732\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    733\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'python'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/anand/.virtualenvs/main/local/lib/python2.7/site-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m   1101\u001b[0m         \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'allow_leading_cols'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex_col\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1102\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1103\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_parser\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTextReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1104\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1105\u001b[0m         \u001b[0;31m# XXX\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/parser.pyx\u001b[0m in \u001b[0;36mpandas.parser.TextReader.__cinit__ (pandas/parser.c:3246)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/parser.pyx\u001b[0m in \u001b[0;36mpandas.parser.TextReader._setup_parser_source (pandas/parser.c:6111)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mIOError\u001b[0m: File /media/nucleoid/raw_data/troy_sandberg/SSW_RNAseq/1_133_S28/2016-10-01_RNA-seq.csv does not exist"
     ]
    }
   ],
   "source": [
    "DF_metadata = pd.DataFrame()\n",
    "for metadata in DF_files.metadata:\n",
    "    df = pd.read_csv(metadata,index_col=0,header=None)\n",
    "    df.loc['metadata'] = metadata\n",
    "    DF_metadata = pd.concat([DF_metadata,df.transpose()])\n",
    "DF_metadata = DF_metadata.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "DF_metadata['carbon-source'] = [re.sub(' |\\(.*\\)','',x.lower()) \n",
    "                                for x in DF_metadata['carbon-source']]\n",
    "DF_metadata['nitrogen-source'] = [re.sub('\\(.*\\)','',x) \n",
    "                                  for x in DF_metadata['nitrogen-source']]\n",
    "DF_metadata['electron-acceptor'] = [re.sub('\\(.*\\)','',x) if type(x)==str else x\n",
    "                                    for x in DF_metadata['electron-acceptor']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "DF_metadata.loc[0,'project'] = 'wt_dhk'\n",
    "DF_metadata.loc[1,'project'] = 'wt_dhk'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "DF_metadata = pd.merge(DF_metadata.drop_duplicates(),DF_files[['condition','rep','metadata']],on='metadata')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for i,row in DF_metadata.iterrows():\n",
    "    DF_metadata.loc[i,'name'] = '__'.join([re.sub(' ','_',row.project.lower()),row.condition,str(row.rep)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DF_metadata.to_csv('metadata.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
